invokeai/app/invocations/upscale_sdx4.py

```diff
@@ -0,0 +1,130 @@
+from typing import Literal, Optional
+
+import torch
+from diffusers import StableDiffusionUpscalePipeline
+from pydantic import Field
+
+from invokeai.app.invocations.baseinvocation import InvocationConfig, InvocationContext
+from invokeai.app.invocations.image import ImageOutput
+from invokeai.app.invocations.latent import TextToLatentsInvocation, get_scheduler
+from invokeai.app.invocations.metadata import CoreMetadata
+from invokeai.app.invocations.model import VaeField
+from invokeai.app.models.image import ImageField, ResourceOrigin, ImageCategory
+from invokeai.app.util.step_callback import stable_diffusion_step_callback
+from invokeai.backend.stable_diffusion import PipelineIntermediateState
+
+
+class UpscaleLatentsInvocation(TextToLatentsInvocation):
+    """Upscales an image using an upscaling diffusion model.
+
+    https://huggingface.co/stabilityai/stable-diffusion-x4-upscaler
+    """
+
+    type: Literal["upscale_sdx4"] = "upscale_sdx4"
+
+    # Inputs
+    image: Optional[ImageField] = Field(description="The image to upscale")
+    vae: VaeField = Field(default=None, description="VAE submodel")
+    metadata: Optional[CoreMetadata] = Field(default=None, description="Optional core metadata to be written to the image")
+    tiled: bool = Field(
+        default=False,
+        description="Decode latents by overlapping tiles(less memory consumption)")
+    # TODO: fp32: bool = Field(DEFAULT_PRECISION=='float32', description="Decode in full precision")
+
+    class Config(InvocationConfig):
+        schema_extra = {
+            "ui": {
+                "title": "Upscale (Stable Diffusion x4)",
+                "tags": ["scale"],
+                "type_hints": {
+                    "model": "model",
+                    "cfg_scale": "number",
+                }
+            }
+        }
+
+    @torch.inference_mode()
+    def invoke(self, context: InvocationContext) -> ImageOutput:
+        image = context.services.images.get_pil_image(self.image.image_name)
+
+        # TODO: are there upscaler LoRA? ControlNet? document if not
+
+        model_manager = context.services.model_manager
+        unet_info = model_manager.get_model(**self.unet.unet.dict(), context=context)
+        vae_info = model_manager.get_model(**self.vae.vae.dict(), context=context)
+
+        # ¿what is this context manager doing?
```

Comment:
the context manages the loading of the model into and out of VRAM

Reply:
oh, I'd looked up the wrong `ModelInfo` class. Now I see the one with the [context manager methods](https://github.com/invoke-ai/InvokeAI/blob/be00a837cce6072ea7bebc93ba4e8ac1cfa81fac/invokeai/backend/model_management/model_manager.py#L275). That makes sense, thanks!